### PR TITLE
[Fields] Migrate store and actions from editor package to fields package

### PR DESCRIPTION
**Pull request by gigitux**
## What?

This PR moves actions and fields to `@wordpress/fields`.

Based on the different types, I created subfolder to collect all the actions that belong at the same type.

- `base-post`: contains all the actions for the `BasePost` type.
- `common` contains all the actions for the `Post` type (that it is an union type).
- `pattern` contains all the actions for the `Pattern` type.

We should change the name of the files inside these folders, but I'm going to do in a dedicated PR.

@youknowriad suggested not to move the store at least for now in `@wordpress/fields` package given that stores are singleton and using them in "bundled packages" create a lot of issues. 

### Old description

<details> 

According to [this feedback](https://github.com/WordPress/gutenberg/pull/65230#pullrequestreview-2296116287), this PR initiates the process of moving the store and certain actions from the `@wordpress/editor` package to the `@wordpress/fields` package. To facilitate easier review and iteration, this work will be divided into multiple smaller PRs. This PR is the first in that series.

### Question

This PR migrates all actions that do not depend on the `@wordpress/editor` package. How should we handle the actions that do depend on the `@wordpress/editor` package?

Some examples:

- `reset-post.tsx`: https://github.com/WordPress/gutenberg/blob/0e671a2eca9c78dfdb45b2149ce9e3fb5bf3ae4a/packages/editor/src/dataviews/actions/reset-post.tsx#L45
- `duplicate-template-part.tsx`: https://github.com/WordPress/gutenberg/blob/0e671a2eca9c78dfdb45b2149ce9e3fb5bf3ae4a/packages/editor/src/dataviews/actions/duplicate-template-part.tsx#L53-L54

Wouldn't it be better if `@wordpress/fields` only exposed the store, allowing actions to be registered based on the post type, while the actual registration occurs on the `@wordpress/editor` side?

I can imagine that for plugins that define new `postType`, the registration will happen on their JS code.

WDYT? @youknowriad @oandregal 

</detail>

### CODE REVIEW

File: packages/editor/src/dataviews/store/private-actions.ts
Comment (gigitux) [inline packages/editor/src/dataviews/store/private-actions.ts:1]: The deleted code has been migrated to the `@wordpress/fields` package

File: packages/fields/src/store/actions.ts
Comment (gigitux) [inline packages/fields/src/store/actions.ts:1]: The commented code has not yet been migrated to the `editor` package. I will complete the migration in a subsequent pull request.

File: packages/fields/src/types.ts
Comment (gigitux) [inline packages/fields/src/types.ts:1]: Copied-pasted from ```packages/editor/src/dataviews/types.ts```

File: packages/fields/src/wordpress-editor.d.ts [hunk 1]
```diff
@@ -0,0 +1 @@
+declare module '@wordpress/editor';
```
Comment (gigitux) [inline packages/fields/src/wordpress-editor.d.ts:1]: This is necessary because `@wordpress/editor` package isn't a TypeScript package

**Comment by youknowriad** [comment]
> This PR migrates all actions that do not depend on the @wordpress/editor package. How should we handle the actions that do depend on the @wordpress/editor package?

Looking at these actions, they don't really depend on the "editor" package, they just revertTemplate and resetTemplate actions that are basically just functions that are misplaced, these functions don't need the editor package at all.

> Wouldn't it be better if @wordpress/fields only exposed the store, allowing actions to be registered based on the post type, while the actual registration occurs on the @wordpress/editor side?

I'm not sure I understand what you're suggestion: field and actions in the "fields" package, the store in the "fields" package but the registration elsewhere?

File: packages/editor/src/components/post-actions/actions.js [hunk 2]
```diff
@@ -14,17 +15,33 @@ export function usePostActions( { postType, onActionPerformed, context } ) {
 	const { defaultActions } = useSelect(
 		( select ) => {
 			const { getEntityActions } = unlock( select( editorStore ) );
+			const { getEntityActions: getEntityActionsFromFieldsPackage } =
+				select( fieldsStore );
```
Comment (youknowriad) [inline packages/editor/src/components/post-actions/actions.js]: One thing that I just thought about is that "stores" are singletons and using them in "bundled packages" create a lot of issues. For instance if a third-party decides to use the package as well, we'll end up with two stores that don't talk to each other.

This makes me thing that maybe a first step here is to extract the fields and the actions but to leave the store and registration in the "editor" package for now.

**Comment by gigitux** [comment]
> Looking at these actions, they don't really depend on the "editor" package, they just revertTemplate and resetTemplate actions that are basically just functions that are misplaced, these functions don't need the editor package at all.
> 
> >

Oh ok! Gotcha! Also, in the case of the `duplicateTemplatePart`, we are using `CreateTemplatePartModalContents` that it is defined in the editor package. 


Sorry for not being clear. I'm going to rephrase/improving my suggestion having new context after you comment.

I'm suggesting to (on long term): 

- fields, actions (that don't depend on `editor` package) and store in the `fields` package.
- keep remaining actions in the `@wordpress/editor` package, registering them via the `registerEntityAction` function.

> This makes me thing that maybe a first step here is to extract the fields and the actions but to leave the store and registration in the "editor" package for now.

👍 It makes sense! I will update the PR!

**Comment by youknowriad** [comment]
> fields, actions (that don't depend on editor package) and store in the fields package.

I don't think there should be fields in Gutenberg that depend on "editor". It's probably just because they were developed there... The CreateTemplate modal can be moved to the "fields" package.

File: packages/editor/src/dataviews/store/private-actions.ts [hunk 3]
```diff
@@ -8,22 +8,24 @@ import { doAction } from '@wordpress/hooks';
...
-import viewPostRevisions from '../actions/view-post-revisions';
-import viewPost from '../actions/view-post';
+import {
+	viewPost,
+	viewPostRevisions,
+	duplicatePost,
+	duplicatePattern,
+	reorderPage,
+	exportPattern,
+	permanentlyDeletePost,
```
Comment (youknowriad) [inline packages/editor/src/dataviews/store/private-actions.ts:26]: If we're going to export actions and fields from the same package, maybe we should consider suffixing these with `Action`?

File: packages/fields/README.md [hunk 4]
```diff
@@ -14,7 +14,53 @@ npm install @wordpress/fields --save
 
 <!-- START TOKEN(Autogenerated API docs) -->
 
-Nothing to document.
+### duplicatePattern
+
+Undocumented declaration.
```
Comment (youknowriad) [inline packages/fields/README.md:19]: Would be cool to add some placeholder JS Docs for these to avoid the "Undocumented declaration". We can definitely do that in a separate PR though.

File: packages/fields/src/actions/base-post/index.ts [hunk 5]
```diff
@@ -0,0 +1,5 @@
+export { default as viewPost } from './view-post';
```
Comment (youknowriad) [inline packages/fields/src/actions/base-post/index.ts:1]: Do we need the "base-post" folder?

File: packages/editor/src/dataviews/store/private-actions.ts (hunk 3 above)
Comment (gigitux) [reply packages/editor/src/dataviews/store/private-actions.ts:26]: Yeah, my idea is to have a structure like this <verb><type>Action. 

https://github.com/WordPress/gutenberg/blob/3616964b435dde324211bfa38838497392ead654/packages/fields/src/actions/base-post/duplicate-post.tsx#L30-L31

For instance, the `duplicatePost` action will be `duplicateBasePostAction`.

File: packages/fields/src/actions/base-post/index.ts (hunk 5 above)
Comment (gigitux) [reply packages/fields/src/actions/base-post/index.ts:1]: I don't have a strong opinion, but I explained why I structured it in this way in the PR description! Happy to revisit it!

Comment (youknowriad) [reply packages/fields/src/actions/base-post/index.ts:1]: I like to start flat personally in general but I don't mind if you prefer it this way.

File: packages/fields/README.md (hunk 4 above)
Comment (gigitux) [reply packages/fields/README.md:19]: After wrapped up the migration, I will:
- add JSDocs
- update file name and action name (as I described [here](https://github.com/WordPress/gutenberg/pull/65261#discussion_r1756829407))

**Review by youknowriad — APPROVED** [review]

**Comment by gigitux** [comment]
I fixed the CI steps regarding mobile version with [ae0efeb](https://github.com/WordPress/gutenberg/pull/65261/commits/ae0efebd5e9c32268499ff40cfd5107d1edc2631) and [f7f1d4d](https://github.com/WordPress/gutenberg/pull/65261/commits/f7f1d4d085bf48c43f2a048f0a5425d3d4ef010b).

This PR is ready to be merged! Thanks for the review! 🙇

File: packages/editor/src/dataviews/store/private-actions.ts (hunk 3 above)
Comment (oandregal) [reply packages/editor/src/dataviews/store/private-actions.ts:26]: I'm a bit confused about why this exports actions. How you all envision the `wordpress/fields` package? If it's about fields, it should only export fields (see [suggestion](https://github.com/WordPress/gutenberg/pull/65196/files#r1756654429)). If it's about dumping every WordPress-related thing we have in DataViews, wouldn't it be best to use a name that communicates that better?

Comment (gigitux) [reply packages/editor/src/dataviews/store/private-actions.ts:26]: I think that the `@wordpress/fields` will expose all the components/actions necessary for WordPress Core - we discussed this [here](https://github.com/WordPress/gutenberg/pull/65230#discussion_r1753668574).
